plugin/pkg/scheduler/algorithm/predicates/predicates.go

```diff
@@ -637,82 +640,115 @@ func NewServiceAffinityPredicate(podLister algorithm.PodLister, serviceLister al
+// 	the match.
+// Otherwise:
+// 	Create an "implicit selector" which gaurantees homogenous labels for all pods, like so.
+// 	Missing labels in the pod: Fill them in by picking a node on the cluster, which already fulfills this svc,
+//	and use that pods labelling information as to fill in any remaining affinity constraints.
+// 	To do this, we "reverse engineer" a selector by introspecting existing pods running under the same service/namespace.
+// 		- L is a label that the ServiceAffinity object needs as a matching constraints.
+// 		- L is not defined in the pod.
+// 		- AND SOME pod from a service is the same namespace, already scheduled onto a node provides that value.
 func (s *ServiceAffinity) CheckServiceAffinity(pod *api.Pod, meta interface{}, nodeInfo *schedulercache.NodeInfo) (bool, []algorithm.PredicateFailureReason, error) {
+	predicateMeta := meta.(*predicateMetadata)
 	node := nodeInfo.Node()
 	if node == nil {
 		return false, nil, fmt.Errorf("node not found")
 	}
-
-	var affinitySelector labels.Selector
-
-	// check if the pod being scheduled has the affinity labels specified in its NodeSelector
-	affinityLabels := map[string]string{}
+	// Check if the pod being scheduled has the affinity labels specified in its NodeSelector
 	nodeSelector := labels.Set(pod.Spec.NodeSelector)
-	labelsExist := true
-	for _, l := range s.labels {
-		if nodeSelector.Has(l) {
-			affinityLabels[l] = nodeSelector.Get(l)
-		} else {
-			// the current pod does not specify all the labels, look in the existing service pods
-			labelsExist = false
+	// Get as many constraints as possible from the pod itself.
+	findLabelsInPod := func(labelsToKeep []string) map[string]string {
+		aL := make(map[string]string)
+		for _, l := range labelsToKeep {
+			if nodeSelector.Has(l) {
+				aL[l] = nodeSelector.Get(l)
+			}
+			// No need to flag missing labels: if the length of this map < labelsToKeep,
+			// we know that we need to introspect nodes to come up with a complet set of
+			// homogenous labelling constraints.
+		}
+		return aL
+	}
+	// Helper function which will allow us to backfill missing constraints with constraints we find in nodes.
+	// Ideally, it won't be used at all if we have all affinity constraints given inside the pod.
+	addUnsetLabelsToMap := func(aL map[string]string, labelsToAdd []string, labelSet labels.Set) {
+		for _, l := range labelsToAdd {
+			// If the pod being scheduled has the label value specified, do not override it
+			if _, exists := aL[l]; exists {
+				continue
+			}
+			// Otherwise
+			if labelSet.Has(l) {
+				aL[l] = labelSet.Get(l)
+			}
+		}
+	}
+	// Helper function to filter pods by namespaces.
+	filterPodsByNamespace := func(pods []*api.Pod) []*api.Pod {
+		filtered := []*api.Pod{}
+		for _, nsPod := range pods {
+			if nsPod.Namespace == pod.Namespace {
+				filtered = append(filtered, nsPod)
+			}
+		}
+		return filtered
+	}
```

Comment:
Do we have a home for queries and data access functions anywhere else...? If so maybe I can just reuse entirely... if not maybe we could start one....